File: backend/src/server/providers/openai-compatible.ts
/**
 * Generic OpenAI-compatible provider adapter.
 *
 * Covers OpenAI, OpenRouter, Moonshot and NVIDIA — all expose
 * OpenAI-shaped /chat/completions endpoints. One adapter + per-provider
 * base URL/credential from the environment keeps the abstraction small.
 */
import {
  ProviderAdapter,
  ProviderChatRequest,
  ProviderChatResult,
  ProviderError,
  estimateTokens,
} from "./types";

type ProviderConfig = { baseUrl: string; apiKeyEnvId: string };

const PROVIDER_CONFIG: Record<string, ProviderConfig> = {
  openai: { baseUrl: "https://api.openai.com/v1", apiKeyEnvId: "OPENAI_API_KEY" },
  openrouter: { baseUrl: "https://openrouter.ai/api/v1", apiKeyEnvId: "OPENROUTER_API_KEY" },
  moonshot: { baseUrl: "https://api.moonshot.cn/v1", apiKeyEnvId: "MOONSHOT_API_KEY" },
  nvidia: { baseUrl: "https://integrate.api.nvidia.com/v1", apiKeyEnvId: "NVIDIA_API_KEY" },
};

export class OpenAICompatibleProvider implements ProviderAdapter {
  constructor(readonly id: string) {}

  get displayName(): string {
    return `OpenAI-compatible: ${this.id}`;
  }

  private get config(): ProviderConfig {
    const cfg = PROVIDER_CONFIG[this.id];
    if (!cfg) throw new ProviderError(`Unknown provider: ${this.id}`, 500, this.id);
    return cfg;
  }

  isConfigured(): boolean {
    return Boolean(process.env[this.config.apiKeyEnvId]);
  }

  async chatCompletion(request: ProviderChatRequest): Promise<ProviderChatResult> {
    const { baseUrl, apiKeyEnvId } = this.config;
    const apiKey = process.env[apiKeyEnvId];
    if (!apiKey) {
      throw new ProviderError(
        `Provider ${this.id} is not configured upstream. TaskFlow has not attached credentials for this provider yet.`,
        503,
        this.id,
      );
    }

    let res: Response;
    const started = Date.now();
    try {
      res = await fetch(`${baseUrl}/chat/completions`, {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${apiKey}`,
          ...(this.id === "openrouter" ? { "HTTP-Referer": "https://taskflow.web-agent.org", "X-Title": "TaskFlow" } : {}),
        },
        body: JSON.stringify({
          model: request.model,
          messages: request.messages,
          temperature: request.temperature,
          ...(request.maxTokens ? { max_tokens: request.maxTokens } : {}),
        }),
        signal: AbortSignal.timeout(60_000),
      });
    } catch (err) {
      throw new ProviderError(
        `Upstream ${this.id} unreachable after ${Date.now() - started}ms`,
        504,
        this.id,
      );
    }

    if (!res.ok) {
      const body = await res.text().catch(() => "");
      throw new ProviderError(
        `Upstream ${this.id} returned ${res.status}`,
        res.status === 429 ? 429 : 502,
        this.id,
      );
    }

    const data = (await res.json()) as {
      choices?: Array<{ message?: { content?: string }; finish_reason?: string }>;
      usage?: { prompt_tokens?: number; completion_tokens?: number };
    };
    const content = data.choices?.[0]?.message?.content ?? "";
    const usage =
      data.usage?.prompt_tokens !== undefined
        ? {
            inputTokens: data.usage.prompt_tokens ?? 0,
            outputTokens: data.usage.completion_tokens ?? 0,
            estimated: false,
          }
        : estimateTokens(request.messages, content);

    return {
      content,
      model: request.model,
      finishReason: data.choices?.[0]?.finish_reason ?? null,
      usage,
      raw: data,
    };
  }
}

File: backend/app/providers/openai_compatible.py
from __future__ import annotations

import time

import httpx

from app.providers.base import ChatCompletionResult, ProviderAdapter, ProviderRequestError

# Bounded retry/backoff for transient failures (429/502/503) — never hammer an
# overloaded upstream provider with immediate retries (see docs/AGENTS.md §58).
RETRYABLE_STATUS = {429, 502, 503, 504}
MAX_RETRIES = 2
BASE_BACKOFF_SECONDS = 0.5


class OpenAICompatibleProvider(ProviderAdapter):
    def __init__(self, name: str, api_base: str, api_key: str):
        self.name = name
        self.api_base = api_base.rstrip("/")
        self.api_key = api_key

    def is_configured(self) -> bool:
        return bool(self.api_key)

    async def health(self) -> bool:
        if not self.is_configured():
            return False
        try:
            async with httpx.AsyncClient(timeout=5) as client:
                resp = await client.get(
                    f"{self.api_base}/models", headers={"Authorization": f"Bearer {self.api_key}"}
                )
                return resp.status_code < 500
        except httpx.HTTPError:
            return False

    async def chat_completion(self, *, model_identifier: str, messages: list[dict], **kwargs) -> ChatCompletionResult:
        if not self.is_configured():
            raise ProviderRequestError(f"Provider '{self.name}' is not configured", status_code=503)

        start = time.monotonic()
        last_error: Exception | None = None

        async with httpx.AsyncClient(timeout=60) as client:
            for attempt in range(MAX_RETRIES + 1):
                try:
                    resp = await client.post(
                        f"{self.api_base}/chat/completions",
                        headers={"Authorization": f"Bearer {self.api_key}"},
                        json={"model": model_identifier, "messages": messages, **kwargs},
                    )
                except httpx.HTTPError as exc:
                    last_error = exc
                    if attempt < MAX_RETRIES:
                        import asyncio

                        await asyncio.sleep(BASE_BACKOFF_SECONDS * (2**attempt))
                        continue
                    raise ProviderRequestError(f"Provider '{self.name}' request failed: {exc}") from exc

                if resp.status_code in RETRYABLE_STATUS and attempt < MAX_RETRIES:
                    import asyncio

                    await asyncio.sleep(BASE_BACKOFF_SECONDS * (2**attempt))
                    continue

                if resp.status_code >= 400:
                    raise ProviderRequestError(
                        f"Provider '{self.name}' returned {resp.status_code}: {resp.text[:500]}",
                        status_code=502,
                    )

                data = resp.json()
                latency_ms = int((time.monotonic() - start) * 1000)
                choice = data["choices"][0]["message"]["content"]
                usage = data.get("usage", {})
                return ChatCompletionResult(
                    content=choice,
                    input_tokens=usage.get("prompt_tokens", 0),
                    output_tokens=usage.get("completion_tokens", 0),
                    raw=data,
                    latency_ms=latency_ms,
                )

        raise ProviderRequestError(f"Provider '{self.name}' request failed after retries: {last_error}")
